### DISCUSSION TITLE
Folder View in List

### DISCUSSION THREAD

**Discussion by zubricks** [payloadcms/payload#4190] [Roadmap]
Design + Build a folder view for all collection types (not just Media collections).

**Comment by PP-Tom** [comment]
We would love this feature, it's the only thing I miss from Storyblok. It's a great way to organise content.

**Reply by TechSynthesis** [reply]
Yup! same here 💯

**Comment by SalahAdDin** [comment]
How is this going on?

**Reply by danpii** [reply]
we would love this feature as well!
@zubricks any update on this?

**Comment by luochuanyuewu** [comment]
Any news on this one?

**Reply by Adaphath** [reply]
There seems to be branch with last commit in October and one of the latest betas integrated.
Seems to be close...

**Reply by SalahAdDin** [reply]
@Adaphath which branch?

**Reply by Adaphath** [reply]
@SalahAdDin https://github.com/payloadcms/payload/tree/feat/beta/folders

**Reply by SalahAdDin** [reply]
> @SalahAdDin https://github.com/payloadcms/payload/tree/feat/beta/folders

Let's check it! 

Thank you

**Comment by nachomglz** [comment]
updates on this?

**Comment by exymax** [comment]
This feature will be a great win, i will start a few new projects with Payload after it's released

**Comment by brightonmike** [comment]
A big, big plus one on this.

**Comment by cgilly2fast** [comment]
Sounds like it will be released soon

**Reply by SalahAdDin** [reply]
Any news about it?

**Reply by immotus** [reply]
Did you hear that sound from a reliable source, e.g., from a core team member? It seems to be hanging in here for years now with little input from the core team 😕

**Reply by cgilly2fast** [reply]
A large commit was pushed today: https://github.com/payloadcms/payload/pull/10030 probably still a couple weeks out

**Comment by rjay-hb** [comment]
@zubricks, is there any update on this feature?

**Reply by zubricks** [reply]
We're so close! We've made a ton of progress and they _will_ be global. You'll be able to create folders for assets, or folders for projects/case studies for example. We posted a few videos on social, but for those of you that aren't check it out below.

https://github.com/user-attachments/assets/a8ea220d-953d-4f53-ac49-2521ef59da5d

https://github.com/user-attachments/assets/94f90fd1-867a-44d3-b7e5-aefef8dd50bd

**Reply by notflip** [reply]
Will the folders have any influence on the filenames (of media?), so does moving a media item inside a folder change it's absolute URL (for example when using with next/image)

**Reply by luochuanyuewu** [reply]
Can't wait to try it out!!!

**Reply by thanhnd1o2** [reply]
Great

**Reply by rxtsel** [reply]
Awesome!!!

**Reply by jmikrut** [reply]
> moving a media item inside a folder change it's absolute URL (for example when using with next/image)

Yes! You will be able to configure prefixes based on folder structure for sure, but this will be optional. You will be able to opt in to that behavior if you want to

**Reply by luochuanyuewu** [reply]
Can't wait to try it out!!! Again

**Comment by alexrah** [comment]
Do you guys have a reliable ETA for this release? Thanks

**Reply by cgilly2fast** [reply]
I have been following the branch [#10030](https://github.com/payloadcms/payload/pull/10030) it looks like they are working out the kinks on merging to main so I am hoping for soon.

**Reply by alexrah** [reply]
yeah me too, the problem with "soon" comes after saying it for months now. As I said, I need a reliable ETA to plan, unfortunately keep saying soon isn't helping. I've been cornered by my CEO who won't accept "soon" as an answer anymore.

**Comment by jmcapra** [comment]
Can't wait to try this feature out

https://x.com/jarrodmflesch/status/1894508882837315933

**Comment by MartinHaun** [comment]
Eagerly awaiting this feature on clients behalf :)

**Comment by jmcapra** [comment]
Folder View has been released! 

https://github.com/payloadcms/payload/releases/tag/v3.39.0

**Comment by hatsumatsu** [comment]
This is amazing! Does this mean we also get the option to switch between list and grid view for regular (non-upload) collections? 🥹

**Reply by SalahAdDin** [reply]
> This is amazing! Does this mean we also get the option to switch between list and grid view for regular (non-upload) collections? 🥹
I guess not.

Did you check it?

**Comment by swheeler7** [comment]
The folders feature is amazing huge thanks to payload team for creating this and just continually shipping such high quality features! Amazing job. I had some time to play around with it and I really love how the ui works and the level of polish feels like google drive or similar which is a huge achievement @JarrodMFlesch 

I do have some comments that I hope can be taken into consideration. The overall implementation is a bit surprising to me and wasn't quite what I was expecting - but not in a bad way at all. When I heard folder view I was basically expecting an alternative to the current list view but which would have been an alternative ui scoped only to the specific collection. What you actually ended up building seems far more powerful than that, however the added complexity does mean that there is some functionality that I was hoping for that is missing, but I think could be added in at some point based on the design pattern you implemented

1) Folder collection scoping - It would be really nice if there was some pattern for specifying collections that folders should be scoped to, rather than all folders being enabled globally. Currently, if I have two folder enabled collections (such as idk - trees and cars for example), and wanted to organize the content of those collections into folders then any folders I create that are specific to the trees collection would also show up in the cars collection and users could end up seeing a "Evergreens" folder in the cars collection. Maybe this feature could be achieved by adding a 'collection-slug' hasMany text field to the 'payload-folders' collection, and when creating folders or editing folders users could choose which collections the folder applies to.

2) UI based access control - This might be a can of worms, but it would be kind of cool to have the ability to limit access to items in folders to specific users, similar to how google drive has folder sharing options. I'm thinking that this access control would only apply in the folder view in the UI and wouldn't replace the stronger code-based access control as it could open up a lot of issues (like if a folder gets deleted who has permissions to move items into specific access protected folders). The new query-presets feature has this functionality, maybe similar logic could be used for folders? 

3) payload-folders collection access control - This could already be an option and maybe I missed it in the docs, but it would be helpful if, when enabling folders in the config, we could add some access control functions to limit who can do what with folders, so that maybe only specific users can create folders, others can just add items to folders.

3) Option to disable the "Browse by Folder" view - while this view and overall pattern of being able to view content from multiple collection in one unified view is really powerfull, it might not be appropriate for all applications.

Again, amazing job with this feature it's definitely a game changer. Currently the only thing that makes me hesitant to use it heavily in my application is the collection scoping. I would push for globally scoped folders to be opt-in vs being the default. The the other features I mention would be nice to have but wouldn't stop me from using the folders and could be added down the line based on the way that the folders are structured

**Reply by SalahAdDin** [reply]
> Option to disable the "Browse by Folder" view - while this view and overall pattern of being able to view content from multiple collection in one unified view is really powerfull, it might not be appropriate for all applications.

Set to `false` to disable it.

**Comment by alexrah** [comment]
I'll check this out pretty soon but for now someone who tried can confirm that this implementation allow to upload assets in sub-folders? I already asked for this few times getting an affirmative response from the team but it doesn't look like so reading the docs.

And from what I remember this folder discussions started from here [Extending Upload to have nested folders](https://github.com/payloadcms/payload/discussions/1961) (closed discussion because marked as duplicate of this one) which is a critical feature for any CMS aiming to host a lot of content.

Besides uploads, I second what said by @swheeler7 especially on folder collection scoping, I didn't expect global folders and honestly I don't have a use case for that atm, it could be cool, but in most cases folders can be useful for a collection and useless for another.

**Reply by inanc-g** [reply]
I can confirm that you can nest folders in the admin UI and it is possible to upload assets into nested folders. 

![image](https://github.com/user-attachments/assets/826db315-00ba-40e4-95c1-2a6df6c7720b)

Note that folders in the admin UI do not affect the structure in s3 storage.  
Moving around objects does not change their location in the bucket.   
(Can't say for local storage, as I only use s3 storage for assets)

**Reply by alexrah** [reply]
with s3 you can define prefixes to organize uploads, that's why this is especially critical for local storage.

**Reply by SalahAdDin** [reply]
> I can confirm that you can nest folders in the admin UI and it is possible to upload assets into nested folders.
> 
> ![image](https://private-user-images.githubusercontent.com/1510491/446801056-826db315-00ba-40e4-95c1-2a6df6c7720b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI1NDA5ODUsIm5iZiI6MTc1MjU0MDY4NSwicGF0aCI6Ii8xNTEwNDkxLzQ0NjgwMTA1Ni04MjZkYjMxNS0wMGJhLTQwZTQtOTVjMS0yYTZkZjZjNzcyMGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MTVUMDA1MTI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzQ2NzY3NmE0ZTdjYWYyOGJjYjg5M2EyYTczNWZlY2U5NzBmOTZkMDQ0MmZjMTg4Mzc4MmU3MmRkNWVhZTE1MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.USkeAn9lzLhR0wzmJ6roF-dEaYKKnUCXQBX3dD--LlE)
> 
> Note that folders in the admin UI do not affect the structure in s3 storage. Moving around objects does not change their location in the bucket. (Can't say for local storage, as I only use s3 storage for assets)

It is working well then, right?

**Reply by jxh0414com** [reply]
what i did was creating an util func get get your current folder ID, and all the parent folders like /users/userId/avatar etc. and have  prefix fields in the media collection and save it that will solve everything. you have prefix for s3 and payloadcms is happy. current im just wondering its its worth doing this or just do multiple collection like product-media, avatar-media etc..

**Reply by SalahAdDin** [reply]
Better use multiple collections, isn't it?
@jxh0414com

**Reply by jxh0414com** [reply]
but the problem with multiple media collection would be you wont be able to reuse the images. right?

**Reply by SalahAdDin** [reply]
no, why?

**Reply by jxh0414com** [reply]
sometime you would want to resuable a image right? actually i dont know if i would for my case. hmmm maybe thats a good idea. but im not sure if this is like a correct way of doing it. im waiting for the new folder update (i think its coming out but there was a breaking bug)

**Reply by SalahAdDin** [reply]
You can chose any image from the folder from any entity.

**Reply by takayumi** [reply]
> Moving around objects does not change their location in the bucket.
> (Can't say for local storage, as I only use s3 storage for assets)

Yeah, but what happens when you move the file to a different folder? How would you update the S3 bucket's prefix then?

**Reply by jxh0414com** [reply]
You would probably need to write some hooks using @aws-sdk/client-s3 package to do some copy command to the new location and delete the old asset. Feel like too much place for bug so I give up on this 


        // 1. Copy the object
        const copyParams = {
          Bucket: destinationBucket,
          CopySource: `${sourceBucket}/${sourceKey}`, // Format: bucketName/objectKey
          Key: destinationKey,
        };
        await s3Client.send(new CopyObjectCommand(copyParams));

        // 2. Delete the original object
        const deleteParams = {
          Bucket: sourceBucket,
          Key: sourceKey,
        };
        await s3Client.send(new DeleteObjectCommand(deleteParams));

**Reply by SalahAdDin** [reply]
> You would probably need to write some hooks using @aws-sdk/client-s3 package to do some copy command to the new location and delete the old asset. Feel like too much place for bug so I give up on this
> 
> ```
>     // 1. Copy the object
>     const copyParams = {
>       Bucket: destinationBucket,
>       CopySource: `${sourceBucket}/${sourceKey}`, // Format: bucketName/objectKey
>       Key: destinationKey,
>     };
>     await s3Client.send(new CopyObjectCommand(copyParams));
> 
>     // 2. Delete the original object
>     const deleteParams = {
>       Bucket: sourceBucket,
>       Key: sourceKey,
>     };
>     await s3Client.send(new DeleteObjectCommand(deleteParams));
> ```

What about the Cloudinary plugin?

**Reply by jxh0414com** [reply]
maybe that too but I haven't really look into it. Not say for sure

**Comment by Aqib-Rime** [comment]
any update on this?
